### PR TITLE
fix(various): Fix #1537, #1538, #1603, #1629, #1630, #1638 and improve DB Upgrade

### LINKED ISSUE #1537: Changing base currency must invalidate all previously defined rates
1. Create new database with **A** currency as a base
2. Download rate history for currencies **B** and **C**
3. Change base currency to **B**
4. Rates for **B** are reseted (=1)
5. ... but rates for **C** are now incorrect!

<hr><small><b>Version</b>: 1.4.0-Alpha.3 • db 7 • git 304a8c27 (2018-04-08) master<br><b>Libs</b>: wxWidgets 3.1.1 (wxGTK 2.24) • wxSQLite3 3.5.9 (SQLite 3.22.0) • Mongoose 6.11 • RapidJSON 1.1.0 • Lua 5.3.3 • libcurl 7.55.1 • OpenSSL 1.0.2g • zlib 1.2.11 • libidn2 2.0.2 • libpsl 0.18.0 • ( libidn2 2.0.2) • librtmp 2.3<br><b>Build</b>: Apr 9 2018 22:34:07 • CMake 3.9.1 • GNU Make 4.1 • gettext 0.19.8.1 • GCC 7.2.0 • Ubuntu 17.10 'artful'<br><b>OS</b>: Ubuntu 17.10 "artful" • Linux 4.13.0-38-generic x86_64 • KDE Polish (UTF-8) • 1280x1024 24bit 96x96ppi</small>

### PR DESCRIPTION
**Pull request by gabriele-v**
Requires https://github.com/moneymanagerex/database/pull/28

---
This change is [Reviewable](https://reviewable.io/reviews/moneymanagerex/moneymanagerex/1635)

### CODE REVIEW

**Comment by GitCop** [comment]
Thanks for your pull request while there were the following issues

 * Commit: 7752b5ecf98bff2dbecc2219f45ddebeb631c789
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

Guidelines are available at https://github.com/moneymanagerex/moneymanagerex/wiki/GitCop

---
This message was auto-generated by https://gitcop.com

**Comment by GitCop** [comment]
Thanks for your pull request while there were the following issues

 * Commit: 7752b5ecf98bff2dbecc2219f45ddebeb631c789
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

Guidelines are available at https://github.com/moneymanagerex/moneymanagerex/wiki/GitCop

---
This message was auto-generated by https://gitcop.com

**Comment by vomikan** [comment]
I can't open any DB file. Assert:
``` HRESULT hr = taskDialogIndirect(&tdc, &msAns, NULL, NULL); ```

**Comment by vomikan** [comment]
New DB. Currencies rates anline update does not working.
image

**Comment by gabriele-v** [comment]
How to reproduce first issue? Simply opening an existing DB? At which version?

Last two errors seems to not be related to changes that I've made, because I haven't changed the main update rates function, anyway I'll try to take a look.

**Comment by vomikan** [comment]
Bug here:
https://github.com/gabriele-v/moneymanagerex/blob/master/src/model/Model_Currency.cpp#L27
or 
https://github.com/gabriele-v/moneymanagerex/blob/master/src/model/Model_Currency.h#L32

**Comment by vomikan** [comment]
I've changed this:
```
enum CURRTYPE { CRYPTO = 0, FIAT};
to     enum CURRTYPE { FIAT = 0, CRYPTO};
```

```
    const auto& crypto_type = Model_Currency::currtype_desc(Model_Currency::CRYPTO);
    auto currencies = Model_Currency::instance().find(Model_Currency::CURRENCY_SYMBOL(base_currency_symbol, NOT_EQUAL));
    for (const auto &currency : currencies)
    {
        if (curr_id > 0 && currency.CURRENCYID != curr_id)
            continue;
        if (curr_id < 0 && !Model_Account::is_used(currency))
            continue;
        const auto& symbol = currency.CURRENCY_SYMBOL.Upper();
        if (symbol.IsEmpty())
            continue;

        if (currency.CURRENCY_TYPE == crypto_type)
            crypto.push_back(symbol);
        else
            fiat.push_back(symbol);
    }
```
It start working

**Comment by vomikan** [comment]
DB for tests

[ttttt2.zip](https://github.com/moneymanagerex/moneymanagerex/files/1977084/ttttt2.zip)

**Comment by vomikan** [comment]
https://github.com/gabriele-v/moneymanagerex/blob/master/src/db/DB_Upgrade.h#L206
should be:
```
        UPDATE CURRENCYFORMATS_V1 SET CURRENCY_TYPE = 'Fiat' WHERE CURRENCY_TYPE <> 'Crypto';
```

**Comment by vomikan** [comment]
for me it stock on 
   ``` db->ReleaseSavepoint("MMEX_Upgrade");```

**Comment by vomikan** [comment]
in my case wxProgressDialog has problem. If commented it's no problems with DB upgrade
```
    //wxProgressDialog progress(_("MMEX database upgrade"), _("Upgrading database..."));
    //progress.Pulse();

    for (; ver < dbLatestVersion; ver++)
    {
        BackupDB(DbFileName, dbUpgrade::BACKUPTYPE::VERSION_UPGRADE, 999, ver);
        if (!UpgradeToVersion(db, ver + 1))
            return false;
    }

    //progress.Destroy();
```

**Comment by gabriele-v** [comment]
Tomorrow I'll try with your DB, but in mine was working fine with the progress dialog. Anyway I'll remove it considering that upgrade are really fast, also slodki one that should be the heaviest.

**Comment by vomikan** [comment]
[test_DB.zip](https://github.com/moneymanagerex/moneymanagerex/files/1977203/test_DB.zip)

**Comment by GitCop** [comment]
Thanks for your pull request while there were the following issues

 * Commit: 7752b5ecf98bff2dbecc2219f45ddebeb631c789
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: 78500c8a537ef5fa3032924cbb78b48dcab42f7b
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

Guidelines are available at https://github.com/moneymanagerex/moneymanagerex/wiki/GitCop

---
This message was auto-generated by https://gitcop.com

**Comment by gabriele-v** [comment]
I've added your fixes and everything it's fine on my machine, could you test it?

File: src/db/DB_Upgrade.h [hunk 1]
```diff
@@ -201,6 +201,12 @@ const std::vector<wxString> dbUpgradeQuery =
         PRAGMA user_version = 10;
     )",
 
+    // Upgrade to version 11
```
Comment (slodki) [inline src/db/DB_Upgrade.h]: Please update database submodule to version taggeg "v11" then regenerate DB_* files

File: src/maincurrencydialog.cpp [hunk 2]
```diff
@@ -150,7 +152,7 @@ void mmMainCurrencyDialog::CreateControls()
     itemBoxSizer22->Add(update_button, g_flagsH);
     update_button->Connect(wxID_STATIC, wxEVT_COMMAND_BUTTON_CLICKED
         , wxCommandEventHandler(mmMainCurrencyDialog::OnOnlineUpdateCurRate), nullptr, this);
-    update_button->SetToolTip(_("Online update currency rate"));
+    update_button->SetToolTip(_("Online update rates for all used currencies"));
```
Comment (slodki) [inline src/maincurrencydialog.cpp]: We should mention that only current (today/yesterday) rates will be updated using this button

File: src/maincurrencydialog.cpp [hunk 3]
```diff
@@ -65,10 +65,11 @@ mmMainCurrencyDialog::mmMainCurrencyDialog(
     bEnableSelect_(bEnableSelect),
     m_static_dialog(false)
 {
-    ColName_[CURR_BASE]   = " ";
+    ColName_[CURR_BASE]   = _("Base");
     ColName_[CURR_SYMBOL] = _("Symbol");
     ColName_[CURR_NAME]   = _("Name");
     ColName_[BASE_RATE]   = _("Last Rate");
+    ColName_[CURR_HIST] = _("Hist.");
```
Comment (slodki) [inline src/maincurrencydialog.cpp]: I think it's not so important to display it here.

File: src/maincurrencydialog.cpp [hunk 4]
```diff
@@ -608,15 +612,12 @@ void mmMainCurrencyDialog::OnHistoryUpdate(wxCommandEvent& /*event*/)
     wxString base_currency_symbol;
     if (!Model_Currency::GetBaseCurrencySymbol(base_currency_symbol))
```
Comment (slodki) [inline src/maincurrencydialog.cpp]: we should replace this with wxASSERT as symbol is now rerequired (UNIQUE NOT NULL)

File: src/maincurrencydialog.cpp [hunk 5]
```diff
@@ -608,15 +612,12 @@ void mmMainCurrencyDialog::OnHistoryUpdate(wxCommandEvent& /*event*/)
...
-            , _("Could not find base currency symbol!")
-            , _("Currency history error"));
+        return mmErrorDialogs::MessageError(this, _("Could not find base currency symbol!"), _("Currency history error"));
     }
 
     wxString msg;
     std::map<wxDateTime, double> historical_rates;
-    bool UpdStatus = GetOnlineHistory(historical_rates
-        , CurrentCurrency->CURRENCY_SYMBOL.Upper(), msg);
+    bool UpdStatus = GetOnlineHistory(historical_rates, CurrentCurrency->CURRENCY_SYMBOL.Upper(), msg);
```
Comment (slodki) [inline src/maincurrencydialog.cpp]: Could we enforce .Upper() at database level or while saving user changes?

File: src/maincurrencydialog.cpp [hunk 6]
```diff
@@ -731,27 +732,22 @@ bool mmMainCurrencyDialog::SetBaseCurrency(int& baseCurrencyID)
...
 
-    Option::instance().BaseCurrency(baseCurrencyID);
-
-    Model_Currency::Data* BaseCurrency = Model_Currency::instance().get(baseCurrencyID);
-    BaseCurrency->BASECONVRATE = 1;
-    Model_Currency::instance().save(BaseCurrency);
-
-    Model_Currency::Data* BaseCurrencyOLD = Model_Currency::instance().get(baseCurrencyOLD);
-    if (BaseCurrencyOLD)
+    if (wxMessageBox(_("Setting a new base currency will delete all static and historic rates:") + "\n"
```
Comment (slodki) [inline src/maincurrencydialog.cpp]: Will this be skipped at db creation time?

**Comment by GitCop** [comment]
Thanks for your pull request while there were the following issues

 * Commit: 7752b5ecf98bff2dbecc2219f45ddebeb631c789
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: 78500c8a537ef5fa3032924cbb78b48dcab42f7b
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

Guidelines are available at https://github.com/moneymanagerex/moneymanagerex/wiki/GitCop

---
This message was auto-generated by https://gitcop.com

**Comment by gabriele-v** [comment]
@slodki: should be fine now.

File: src/maincurrencydialog.cpp (hunk 4 above)
Comment (slodki) [reply src/maincurrencydialog.cpp]: I thought about this:
```c++
wxASSERT_MSG(Model_Currency::GetBaseCurrencySymbol(base_currency_symbol), "Could not find base currency symbol");
```
replacing all `if() { }`

**Comment by GitCop** [comment]
Thanks for your pull request while there were the following issues

 * Commit: 7752b5ecf98bff2dbecc2219f45ddebeb631c789
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: 78500c8a537ef5fa3032924cbb78b48dcab42f7b
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

Guidelines are available at https://github.com/moneymanagerex/moneymanagerex/wiki/GitCop

---
This message was auto-generated by https://gitcop.com

**Comment by GitCop** [comment]
Thanks for your pull request while there were the following issues

 * Commit: 7752b5ecf98bff2dbecc2219f45ddebeb631c789
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: 78500c8a537ef5fa3032924cbb78b48dcab42f7b
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

Guidelines are available at https://github.com/moneymanagerex/moneymanagerex/wiki/GitCop

---
This message was auto-generated by https://gitcop.com

**Pull request merged**